---
title: "MiFish 2022 Bioinformatics for Stormies"
author: "Gemma Clucas"
date: '2022-12-02'
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1. Import the data into Qiime2

First, load qiime environment and cd to correct directory in the terminal. 
```
cd /Users/gemmaclucas/GitHub/Fecal_metabarcoding/Stormies-Keenan-Yakola
conda activate qiime2-2021.4
```

The raw reads are saved on my solid state hardrive, Data_SS1. There are nearly two whole plates, although the second plate has a few Troppy samples on it and also some BLGU samples. I will leave them in for now.

```
qiime tools import\
  --type 'SampleData[PairedEndSequencesWithQuality]'\
  --input-path /Volumes/Data_SS1/MiFish/LESP_KeenanYakola_2022/Plate23/reads \
  --input-format CasavaOneEightSingleLanePerSampleDirFmt\
  --output-path MiFish_2022/demux_plate1.qza
  
qiime tools import\
  --type 'SampleData[PairedEndSequencesWithQuality]'\
  --input-path /Volumes/Data_SS1/MiFish/LESP_KeenanYakola_2022/Plate24/reads \
  --input-format CasavaOneEightSingleLanePerSampleDirFmt\
  --output-path MiFish_2022/demux_plate2.qza
``` 

Summarise read quality and number of reads for each plate. 
```
cd MiFish_2022/

for K in {1..2}; do
  qiime demux summarize \
    --i-data demux_Plate$K.qza \
    --o-visualization demux_Plate$K.qzv
done
```
Plate 23 seems to have high and fairly even read numbers. Plate 24 seems less even but with higher quality potentially.

## 2. Trim primers using cutadapt

The MiFish sequences are:   

F primer: GTCGGTAAAACTCGTGCCAGC (21 bp)   
R primer: CATAGTGGGGTATCTAATCCCAGTTTG (27 bp)   

### Trim 3' ends first
At the 3' end of the read, the primer will have been read through after reading the MiFish region. I need to be looking for the reverse complement of the reverse primer in read 1 (—p-adapter-f) and the reverse complement of the forward primer in R2 (—p-adapter-r).

F primer reverse complement: GCTGGCACGAGTTTTACCGAC   
R primer reverse complement: CAAACTGGGATTAGATACCCCACTATG    

```
for K in {1..2}; do
  qiime cutadapt trim-paired \
    --i-demultiplexed-sequences demux_Plate$K.qza \
    --p-adapter-f CAAACTGGGATTAGATACCCCACTATG \
    --p-adapter-r GCTGGCACGAGTTTTACCGAC \
    --o-trimmed-sequences trimd_Plate$K.qza \
    --verbose > cutadapt_out_Plate$K.txt
done
```
To see how much data passed the filter for each sample:
```
grep "Total written (filtered):" cutadapt_out_Plate1.txt 
grep "Total written (filtered):" cutadapt_out_Plate2.txt
```
77% is the normal amount to see here. Lots of samples on plate 23 are around 77%, and slightly fewer on plate 24, but this might be all the troppy samples perhaps.

Make new visualisations to see how many sequences are left and their quality scores.
```
for K in {1..2}; do
  qiime demux summarize \
    --i-data trimd_Plate$K.qza \
    --o-visualization trimd_Plate$K.qzv
done 
```
This looks good and has cleaned up some of the lower quality scores that were on plate 23.

### Trim 5' ends of reads
All R1 should begin with the forward primer: GTCGGTAAAACTCGTGCCAGC (21 bases).
All R2 should begin with the reverse primer: CATAGTGGGGTATCTAATCCCAGTTTG (27 bases).

Trim these with the following commands:

```
for K in {1..2}; do
  qiime cutadapt trim-paired \
    --i-demultiplexed-sequences trimd_Plate$K.qza \
    --p-front-f GTCGGTAAAACTCGTGCCAGC \
    --p-front-r CATAGTGGGGTATCTAATCCCAGTTTG \
    --o-trimmed-sequences trimd2_Plate$K.qza \
    --verbose > cutadapt_out2_Plate$K.txt
done
```

To see how much data passed the filter for each sample:
```
grep "Total written (filtered):" cutadapt_out2_Plate1.txt 
grep "Total written (filtered):" cutadapt_out2_Plate2.txt
```
About 88% passed the filter here for all samples. That's pretty normal.


## 3. Denoise with dada2
I am going to use the same settings that I found worked best for the 2017-2019 tern fecal samples, except I am adding the --p-min-overlap parameter. If I don't use this, I seem to be getting a load of rubbish reads which are 250bp long and start with long strings of Cs. This option is only available in qiime2-2021.4 and later. I didn't get these rubbish reads before, so I'm not sure what has changed, but the overlap filter seems to fix it. 

Note, this step can be a little slow to run.
```
for K in {1..2}; do
  qiime dada2 denoise-paired \
    --i-demultiplexed-seqs trimd2_Plate$K.qza \
    --p-trunc-len-f 133 \
    --p-trunc-len-r 138 \
    --p-trim-left-f 0 \
    --p-trim-left-r 0 \
    --p-min-overlap 50 \
    --p-n-threads 16 \
    --o-representative-sequences rep-seqs_Plate$K \
    --o-table table_Plate$K \
    --o-denoising-stats denoise_Plate$K
done
```

Create visualizations for the denoising stats.
```
for K in {1..2}; do  
  qiime metadata tabulate\
    --m-input-file denoise_Plate$K.qza\
    --o-visualization denoise_Plate$K.qzv
done
```
These all look good.

## 4. Merge across plates

Note that this requires metadata for each sample, contained in metadata.txt. 

```
qiime feature-table merge \
  --i-tables table_Plate1.qza \
  --i-tables table_Plate2.qza \
  --p-overlap-method sum \
  --o-merged-table merged-table.qza
  
qiime feature-table summarize \
    --i-table merged-table.qza \
    --m-sample-metadata-file metadata.txt \
    --o-visualization merged-table
    
qiime feature-table merge-seqs \
  --i-data rep-seqs_Plate1.qza \
  --i-data rep-seqs_Plate2.qza \
  --o-merged-data merged_rep-seqs.qza
  
qiime feature-table tabulate-seqs \
  --i-data merged_rep-seqs.qza \
  --o-visualization merged_rep-seqs.qzv

```

## 5. Assign taxonomy using same database as for 2021 samples
Note, I am using the same database files that I used for Will's puffin samples. This requires an older version of Qiime2.

```
conda activate qiime2-2019.4

./mktaxa.py \
  ncbi-refseqs-withHuman.qza \
  ncbi-taxonomy-withHuman.qza \
  merged_rep-seqs.qza
```
Make the visualisation.
```
qiime metadata tabulate \
  --m-input-file superblast_taxonomy.qza \
  --o-visualization superblast_taxonomy
```

## 6. Make some barplots
I want to see how many stormy (these are coming up as penguin), human, and unassigned sequences there were.
```
qiime taxa barplot \
  --i-table merged-table.qza \
  --i-taxonomy superblast_taxonomy.qza \
  --m-metadata-file metadata.txt \
  --o-visualization barplot_before_filtering.qzv
```
It looks like a lot of the early samples that were collected at Matinicus Rock have a lot of bird DNA in them. These were from incubating birds, so this is likely why - they hadn't eaten in a while. Some of the extraction blanks have human DNA in them, less than 100 reads though.

More worryingly a couple of the extraction blanks have quite a lot of fish DNA in them on plate 24. See how much of this gets through the rarefaction.


## 7. Remove non-food reads
Filter out any sequences from the bird, mammals (human), and unnassigned sequences since we're not interested in these.
```
qiime taxa filter-table \
  --i-table merged-table.qza \
  --i-taxonomy superblast_taxonomy.qza \
  --p-exclude Unassigned,Aves,Mammalia \
  --o-filtered-table merged_table_noBirdsMammalsUnassigned.qza
  
qiime feature-table summarize \
    --i-table merged_table_noBirdsMammalsUnassigned.qza \
    --m-sample-metadata-file metadata.txt \
    --o-visualization merged_table_noBirdsMammalsUnassigned
```

Barplot having removed bird/human/unassigned DNA:
```
qiime taxa barplot \
  --i-table merged_table_noBirdsMammalsUnassigned.qza \
  --i-taxonomy superblast_taxonomy.qza \
  --m-metadata-file metadata.txt \
  --o-visualization barplot_noBirdsMammalsUnassigned.qzv
```

## 9. Calculate alpha rarefaction curves 

Reload qiime2-2021.4 first.
```
conda activate qiime2-2021.4

qiime taxa collapse \
  --i-table merged_table_noBirdsMammalsUnassigned.qza \
  --i-taxonomy superblast_taxonomy.qza \
  --p-level 7 \
  --o-collapsed-table merged_table_noBirdsMammalsUnassigned_collapsed.qza

qiime diversity alpha-rarefaction \
  --i-table merged_table_noBirdsMammalsUnassigned_collapsed.qza \
  --m-metadata-file metadata.txt \
  --p-min-depth 100 \
  --p-max-depth 10000 \
  --o-visualization alpha-rarefaction-100-10000
```
The curve for the observed OTUs (species) increases from 100 to 4000, and again above 15000, which is very high - maybe we're picking up super low concentrations of DNA from past meals at this depth. 

Repeat the rarefaction for values less than 4000.
```
qiime diversity alpha-rarefaction \
  --i-table merged_table_noBirdsMammalsUnassigned_collapsed.qza \
  --m-metadata-file metadata.txt \
  --p-min-depth 100 \
  --p-max-depth 4000 \
  --o-visualization alpha-rarefaction-100-4000
```
Shannon diversity is still flat for the samples. The observed OTUs plateaus at five features by 500 and then increases to six again by 2500. They are obviously eating much more diverse diets than terns, so I will use 2500 as the rarefaction depth.

Note: I'm just looking at the LESP samples here. The troppy samples all have only one species (most commonly saury) and so could have a much lower depth (100 reads?). Similarly for the black guillemots, they plateau around 1500 sequences, so could be rarefied to a lower depth than the storm petrel samples.



## 10. Split out LESP samples and rarefy to a deph of 2500 and redo barplots

Split out the storm petrel samples only.
```
qiime feature-table filter-samples \
  --i-table merged_table_noBirdsMammalsUnassigned.qza \
  --m-metadata-file metadata.txt \
  --p-where "Species='LESP' OR Species='MOCK' OR Species='PCR_BLANK' \
            OR Species='EXT_BLANK' OR Species='FLD_BLANK'" \
  --o-filtered-table LESP_table_noBirdsMammalsUnassigned.qza \
  --verbose
  
qiime feature-table filter-seqs \
  --i-data merged_rep-seqs.qza \
  --i-table LESP_table_noBirdsMammalsUnassigned.qza \
  --o-filtered-data LESP_seqs.qza
```


Rarefy. Note, this is done on the un-collapsed table.
```
qiime feature-table rarefy \
  --i-table LESP_table_noBirdsMammalsUnassigned.qza \
  --p-sampling-depth 2500 \
  --o-rarefied-table LESP_table_noBirdsMammalsUnassigned_rarefied2500
  
qiime taxa barplot \
  --i-table LESP_table_noBirdsMammalsUnassigned_rarefied2500.qza \
  --i-taxonomy superblast_taxonomy.qza \
  --m-metadata-file metadata.txt \
  --o-visualization barplot_LESP_noBirdsMammalsUnassigned_rarefied2500.qzv
```

There seem to be pairs and even sets of three samples that resemble each other way more than I would expect by chance. I have coloured these on the plate layout and sample spreadsheets. 

According to where they occur on the plates i.e. not always right next to each other, I do not think this happened during the extraction or PCR set-up. More evidence for this is that all the extraction blanks are blank, except one, and the one that is contaminated does not exactly match another sample nearby it on the plate, so I think it was some random cross-contamination that happened in that blank.

Given that the samples are close in numbers but not always next to each other on the plate, this means they either got cross-contaminated during collection or shipping.

A lot of them were collected one after the other on the same day, but one pair ```LESP_MAT_AD_2022_065``` and ```LESP_MAT_AD_2022_066``` were collected on different days from different substrates, suggesting that this is not due to contamination during collections.

Therefore, is it because there was a lot of leakage when they arrived? It does look like a lot of the affected samples were pairs where one of them had leaked (i.e. it is still visible in the threads of the lids) so I do think that this is the most likely explanation, as opposed to it happening during the sampling itself. What a bummer!

What can we do? Just report the list of species detected I guess. I don't think we can rely on the frequency of occurrence data as it's impossible to know how many other samples may have been affected across the plates.




